Akka.Serialization.MessagePack/MsgPackSerializer.cs

```diff
@@ -39,35 +34,118 @@ public MsgPackSerializer(ExtendedActorSystem system, Config config)
+                // - Is there a Ctor that will take ActorSystem/ExtendedActorSystem?
+                // - Is there a Static 'Instance' Property/Field?
+                // - Is there a Public, Parameterless Ctor?
+                
+                var ctors = type.GetConstructors();
+                var actorSystemCtorMaybe = ctors.FirstOrDefault(r =>
+                {
+                    var p = r.GetParameters();
+                    if (p.Length == 1 && p[0].ParameterType
+                            .IsAssignableFrom(typeof(ExtendedActorSystem)))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                });
+                if (actorSystemCtorMaybe != null)
+                {
+                    return (IFormatterResolver)actorSystemCtorMaybe.Invoke(new[]
+                        { system });
+                }
+                
+                var props = type.GetProperties(BindingFlags.Static | BindingFlags.Public);
+                foreach (var propertyInfo in props)
+                {
+                    if (propertyInfo.Name == "Instance")
+                    {
+                        return (IFormatterResolver)propertyInfo.GetValue(null);
+                    }
+                }
+                var fields = type.GetFields(BindingFlags.Static | BindingFlags.Public);
+                foreach (var fieldInfo in fields)
+                {
+                    if (fieldInfo.Name == "Instance")
+                    {
+                        return (IFormatterResolver)fieldInfo.GetValue(null);
+                    }
+                }
+
+                var defaultCtor =
+                    ctors.FirstOrDefault(r => r.GetParameters().Length == 0);
+                if (defaultCtor != null)
+                {
+                    return (IFormatterResolver)defaultCtor.Invoke(Array.Empty<object>());
+                }
+
+                throw new ArgumentException(
+                    $"Type {type} does not contain a static 'Instance' Property/Field, Ctor that takes ActorSystem, or Parameterless Ctor!");
             }
             else
             {
-                return MessagePackSerializer.NonGeneric.Serialize(obj.GetType(), obj);
+                throw new ArgumentException(
+                    $"Type {type} is not assignable to IMessageFormatter!");
             }
         }
 
-        public override object FromBinary(byte[] bytes, Type type)
+        public MsgPackSerializer(ExtendedActorSystem system,
+            MsgPackSerializerSettings settings) : base(system)
         {
-            if (_settings.EnableLz4Compression)
+            _settings = settings;
+            _resolver = CompositeResolver.Create(_settings.OverrideConverters
+                .Select(t => LoadFormatterResolverByType(t, system))
+                .Concat(new[]
+                {
+                    SerializableResolver.Instance,
```

Comment:
Got it... so resolvers are akin to type converters or whatever in Newtonsoft.Json - special case handling?